tests/test_edgeql_expressions.py

```diff
@@ -906,189 +925,28 @@ class TestExpressions(tb.QueryTestCase):
-        await self.assert_query_result("""
-            SELECT json_typeof(<json>'2');
-            SELECT json_typeof(<json>'2');
-            SELECT json_typeof(<json>'"foo"');
-            SELECT json_typeof(<json>'true');
-            SELECT json_typeof(<json>'false');
-            SELECT json_typeof(<json>'null');
-            SELECT json_typeof(<json>'[]');
-            SELECT json_typeof(<json>'[2]');
-            SELECT json_typeof(<json>'{}');
-            SELECT json_typeof(<json>'{"a": 2}');
-        """, [
-            ['number'],
-            ['number'],
-            ['string'],
-            ['boolean'],
-            ['boolean'],
-            ['null'],
-            ['array'],
-            ['array'],
-            ['object'],
-            ['object'],
-        ])
-
-    @unittest.expectedFailure
-    async def test_edgeql_expr_json_08(self):
-        await self.assert_query_result("""
-            SELECT json_array_unpack(<json>'[1, "a", null]');
-        """, [
-            [1, 'a', None],
-        ])
-
-    @unittest.expectedFailure
-    async def test_edgeql_expr_json_09(self):
-        await self.assert_query_result("""
-            SELECT json_object_unpack(<json>'{"q": 1, "w": "a", "e": null}');
-        """, [
-            {
-                ('q', 1),
-                ('w', 'a'),
-                ('e', None),
-            },
-        ])
-
-    @unittest.expectedFailure
-    async def test_edgeql_expr_json_10(self):
-        await self.assert_query_result("""
-            SELECT <json>'[1, 'a', 3]'[0] = <json>1;
-            SELECT <json>'[1, 'a', 3]'[1] = <json>'a';
-            SELECT <json>'[1, 'a', 3]'[2] = <json>3;
-            SELECT <json>'[1, 'a', 3]'[3] ?= <json>{};
-        """, [
-            [True],
-            [True],
-            [True],
-            [True],
-        ])
+    async def test_edgeql_expr_string_03(self):
+        with self.assertRaisesRegex(
+                # FIXME: possibly a different error should be used here
+                exc.UnknownEdgeDBError,
+                r'string index 10 is out of bounds'):
+            await self.con.execute("""
+                SELECT '123'[10];
+            """)
 
-    @unittest.expectedFailure
-    async def test_edgeql_expr_json_11(self):
-        await self.assert_query_result("""
-            SELECT <json>'{"a": 1, "b": null}'["a"] = <json>1;
-            SELECT <json>'{"a": 1, "b": null}'["b"] = <json>'null';
-            SELECT <json>'{"a": 1, "b": null}'["c"] ?= <json>{};
-        """, [
-            [True],
-            [True],
-            [True],
-        ])
+    async def test_edgeql_expr_string_04(self):
+        # There's an inherent problem with negative offsets that they
+        # resolve into some other offset that gets passed to substr.
+        # So potentially this can be prettified by introducing an
+        # extra layer of a function call or some other mechanism for
+        # storing the "original" offset value.
```

Comment:
In this case it's better to give _no_ index at all than to give a misleading value.

Reply:
There are a number of issues with indexes and slices in general that I'll be addressing in a separate PR. These issues are not JSON-specific.